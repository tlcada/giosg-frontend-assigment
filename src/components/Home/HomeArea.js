import React, { Component } from 'react';
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { withStyles, TextField, Button, SnackbarContent } from '@material-ui/core';

import { DateField } from "../Field";
import ReportingView from './components/Reporting/ReportingView';
import { fetchReporting } from "./components/Reporting/action";
import { REPORTING } from '../../mock/reporting'
import config from '../../config/config';

const styles = theme => ({
    container: {
        display: 'flex',
        flexWrap: 'wrap'
    },
    textField: {
        marginLeft: theme.spacing.unit,
        marginRight: theme.spacing.unit,
        width: 200
    },
    button: {
        margin: theme.spacing.unit,
    },
    snackbar: {
        margin: theme.spacing.unit,
    }
});

class Home extends Component {

    constructor(props) {
        super(props);
        this.state = {
            startDate: this.getLocalStorageValue('startDate', "2017-05-01"),
            endDate: this.getLocalStorageValue('endDate', "2017-06-15"),
            accessToken: this.getLocalStorageValue('accessToken', ""),
            error: "",
            showReport: false
        };

        this.onChangeValue = this.onChangeValue.bind(this);
    }

    getLocalStorageValue(key, defaultVal) {
        const value = localStorage.getItem(key);
        return value === null ? defaultVal : value;
    }

    onChangeValue(e) {
        this.setState({
            [e.target.name]: e.target.value
        });
        localStorage.setItem(e.target.name, e.target.value);
    }

    showReport() {
        // Make simple validation
        if (this.state.startDate === "" || this.state.endDate === "" || this.state.accessToken === "") {
            this.setState({ error: "Fill in all the fields."});
        } else {
            this.setState({ error: ""});
            if (!config.useDummyData) {
                this.props.fetchReporting(this.state.startDate, this.state.endDate, this.state.accessToken);
            }
        }

        this.setState({ showReport: true});
    }

    renderReport() {
        if (this.state.showReport) {
            let reportingData;
            if (config.useDummyData) {
                reportingData = { error: "", reporting: REPORTING };
            } else {
                const reporting = this.props.reportingObj;
                reportingData = { error: reporting.error, reporting: reporting.reporting };
            }

            if (Object.keys(reportingData.reporting).length && this.state.error === "") {
                return (
                    <ReportingView reporting={reportingData.reporting} />
                );
            } else {
                let errorMsg = (this.state.error !== "") ? this.state.error : reportingData.error;
                return(
                    <SnackbarContent className={this.props.classes.snackbar} message={"Error: " + errorMsg} />
                )
            }
        } else {
            return(
                <div></div>
            )
        }
    }

    render() {
        const classes = this.props.classes;

        return(
            <div>
                <form className={this.props.classes.container}>
                    <DateField
                        name="startDate"
                        label="Start date"
                        className={classes.textField}
                        defaultValue={this.state.startDate}
                        onChange={this.onChangeValue}
                    />
                    <DateField
                        name="endDate"
                        label="End date"
                        className={classes.textField}
                        defaultValue={this.state.endDate}
                        onChange={this.onChangeValue}
                    />
                    <TextField
                        name="accessToken"
                        label="Access token"
                        className={this.props.classes.textField}
                        margin="normal"
                        required
                        InputLabelProps={{
                            shrink: true
                        }}
                        onChange={this.onChangeValue}
                        value={this.state.accessToken}
                    />
                    <Button
                        variant="outlined" onClick={() => this.showReport()} size="small" color="primary" className={classes.button}>
                        Show Results
                    </Button>
                </form>

                {this.renderReport()}
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        reportingObj: state.reportingReducer
    }
};

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({fetchReporting: fetchReporting}, dispatch)
};

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Home));
