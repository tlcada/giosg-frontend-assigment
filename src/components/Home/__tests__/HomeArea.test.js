import React from 'react';
import { shallow } from 'enzyme';
import HomeArea from '../HomeArea';
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

describe('home', () => {
    let store;

    beforeEach(() => {
        const mockStore = configureMockStore([thunk]);
        store = mockStore({});
    });

    it('renders without crashing', () => {
        const wrapper = shallow(<HomeArea store={store} />).dive();
        expect(wrapper).toMatchSnapshot();
    });
});
