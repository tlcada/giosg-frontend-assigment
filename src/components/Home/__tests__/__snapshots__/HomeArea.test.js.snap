// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Home)
    fetchReporting={[Function]}
    reportingObj={undefined}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "classes": Object {
        "button": "Home-button-03",
        "container": "Home-container-01",
        "snackbar": "Home-snackbar-04",
        "textField": "Home-textField-02",
      },
      "fetchReporting": [Function],
      "reportingObj": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "classes": Object {
          "button": "Home-button-03",
          "container": "Home-container-01",
          "snackbar": "Home-snackbar-04",
          "textField": "Home-textField-02",
        },
        "fetchReporting": [Function],
        "reportingObj": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
